### ISSUE TITLE
Question: Alerts Not Working

### ISSUE THREAD

**Issue by jgerschk** [influxdb/kapacitor#105]
First, this could be a new user issue more than anything; however, I've looked over everything and don't seem to see why this is not being triggered.

stream
    .from().measurement('metric')
    .where(lambda: "metric" == 'UR')
    .groupBy('symbol')
    .window()
        .period(5s)
    .alert()
        .id('kapacitor/{{index .Tags "symbol" }}')
        .message('{{index .Fields "tick" }}')
        .crit(lambda: "tick" > 1)
        .log('/tmp/alerts.log')

from kapacitor list tasks

users                         stream    true      true      ["kapacitor"."default"]

I do see it enabled.

from kapacitor show users

Name: users
Error:
Type: stream
Enabled: true
Executing: true
Databases Retention Policies: ["kapacitor"."default"]
TICKscript:
stream
    .from().measurement('metric')
    .where(lambda: "metric" == 'UR')
    .groupBy('symbol')
    .window()
        .period(5s)
    .alert()
        .id('kapacitor/{{index .Tags "symbol" }}')
        .message('{{index .Fields "tick" }}')
        .crit(lambda: "tick" > 1)
        .log('/tmp/alerts.log')

DOT:
digraph users {
stream0 -> stream1;
stream1 -> groupby2;
groupby2 -> window3;
window3 -> alert4;
}

And yes, that is not a typo, the measurement in influx is called metric and we have tags called metric.  our field is not called value it's called tick (I don't think that is the issue as I see other examples that use different field names).  Also, the kapacitor logs do show the stream coming in, and within influxdb I do see the subscription.

Any suggestions on what the issue is between this user it the keyboard? :)

Thanks.

**Comment by nathanielc** [comment]
At a quick glance it looks like you are missing the every property on the
window node. We should probably add some validation logic for that, but the
window is likely not emitting any data.

**Comment by jgerschk** [comment]
What if I omit the entire .window section?  Would that matter, or would that cause it to not work also?  I've tried it several ways.

Name: users
Error:
Type: stream
Enabled: true
Executing: true
Databases Retention Policies: ["kapacitor"."default"]
TICKscript:
stream
    .from().measurement('metric')
    .where(lambda: 'metric' == 'UR')
    .groupBy('symbol')
    .alert()
        .id('kapacitor/{{index .Tags "symbol" }}')
        .message('{{index .Fields "tick" }}')
        .crit(lambda: 'tick' > 1)
        .log('/tmp/alerts.log')


DOT:
digraph users {
stream0 -> stream1;
stream1 -> groupby2;
groupby2 -> alert3;
}

P.S.  Thanks for getting back to me.

**Comment by nathanielc** [comment]
Looking at this with fresh eyes this morning I am seeing a few more errors in the script. In general I would recommend simplifying the script to a working state and then incrementally add from there. Using the record/replay workflow should make that easier.

The errors I see are that you are using single quotes `'` when you probably meant double quotes `"`. In lambda expressions single quotes are string literals and double quotes are identifiers to tags or fields. The where  lambda expression is `lambda: 'metric' == 'UR'` which is always false since the literal `metric` does not equal the literal `UR`. Looks like you may have a similar issue in the crit lambda expression too.

**Comment by jgerschk** [comment]
I'll get back on this later this afternoon (US time); however a few quick comments.

Unfortunately, this may very well be issues between me and understanding; however, if I understand you correctly .where(lambda:'metric' == 'UR') will not work.  however, the where, if switch to double quotes would?  Confirm if I understand what you are saying to be correct.  This would apply to the .crit section as well, changing to "tick" (double quotes would)?

**Comment by nathanielc** [comment]
Yep.

Change the where to: `.where(lambda: "metric" == 'UR')`.
Change the crit to: `.crit(lambda: "tick" > 1)`.

**Comment by jgerschk** [comment]
That's what I thought.  So I went back and double check my work again (below) -- yet, nothing is ever written out to the log, or the log for kapacitor indicating and error.  I've confirmed again that I see stream data, and confirmed there is data within influx stream for that metric value.  Question, could difference between UTC and local time be the issue?  Or is that meaningless?  

Name: users
Error:
Type: stream
Enabled: true
Executing: true
Databases Retention Policies: ["kapacitor"."default"]
TICKscript:
stream
    .from().measurement('metric')
    .where(lambda: "metric" == 'UR')
    .alert()
        .crit(lambda: "tick" > 1)
        .log('/tmp/alerts.log')


DOT:
digraph users {
stream0 -> stream1;
stream1 -> alert2;
}

**Comment by nathanielc** [comment]
At this points I would try short circuiting the logic to make sure its not an issue with the logic. Add an `OR True` to each lambda expression so that is always evals to true.

Are you using the record/replay workflow? If not I'd recommend it as you can verify data was captured by the recording using `kapacitor list recordings` and noting the size of the recording is more than a few bytes.

If you still can't get it to work try setting the log level to debug or digging into throughput numbers so you can see where data gets dropped. See [this](https://groups.google.com/forum/?pli=1#!topic/influxdb/K1tWkAKamTA)  for explanation for how to check throughput numbers.

**Comment by jgerschk** [comment]
That is worth a try, setting it to OR True.

I did at first use the record/replay but that did not seem to work so I went this approach.  How, I still have those sets saved, size is smaller than I would have thought for the data coming in.

Is there a way to see the "raw" data in those records?

P.S.  Thanks for all the help here.

**Comment by nathanielc** [comment]
The srpl files are just gzipped, so you can take a look at their contents
that way. Basically they are a slightly modified line protocol format.

On Thu, Dec 17, 2015 at 1:23 PM, jgerschk <notifications@github.com> wrote:

> That is worth a try, setting it to OR True.
>
> I did at first use the record/replay but that did not seem to work so I
> went this approach. How, I still have those sets saved, size is smaller
> than I would have thought for the data coming in.
>
> Is there a way to see the "raw" data in those records?
>
> P.S. Thanks for all the help here.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/influxdb/kapacitor/issues/105#issuecomment-165572572>.
>



-- 
-Nathaniel Cook

**Comment by jgerschk** [comment]
trying to keep it simple.

stream
    .from().measurement('metric')
    .where(lambda: "metric" == 'UR' OR TRUE)
    .alert()
        .log('/tmp/alerts.log')

doing a record yields no data in the replay files.

I'm open for suggestions at this point.

**Comment by jgerschk** [comment]
I'm starting to think this may be and influx issue -- did a tcpdump based on the subscription port, and I don't see any data being transmitted over the write.  That could be why these aren't working.

**Comment by jgerschk** [comment]
Did a little test, tcpdump of the data.  And from the capture I see metrics that should have matched but did not based on 


Name: users
Error:
Type: stream
Enabled: true
Executing: true
Databases Retention Policies: ["kapacitor"."default"]
TICKscript:
stream
.from().measurement('metric')
.where(lambda: "metric" == 'UR')
.alert()
.crit(lambda: "tick" > 1)
.log('/tmp/alerts.log')

DOT:
digraph users {
stream0 -> stream1;
stream1 -> alert2;

**Comment by jgerschk** [comment]
I just want to share my experience and results here.  I finally got this working.  There was nothing wrong with my tick scripts above other than what @nathanielc pointed out first.

the name issue was the define statement.  I was miss understanding the -dbrp option and using a DB that did not exist.  Once I changed that to the DB/RP that should be worked against this all started to work.

So basically what I did to start was this

kapacitor define -name alert - type stream - tick alert.tick -dbrp kapacitor

Well kapacitor was not the DB the data was hosed in and that was the root of the problem. 

I'm closing this.
